src/store/themeModule.js

```diff
@@ -2,14 +2,41 @@ import { SAVE_PREFERRED_THEME } from './mutationTypes';
 
 export default {
   state: {
-    preferredTheme: null,
+    preferredTheme: {
+      fromOs: null,
+      fromLocalStorage: null,
+      fromToggle: null,
+    },
   },
   mutations: {
-    [SAVE_PREFERRED_THEME](state, preferredTheme) {
-      state.preferredTheme = preferredTheme;
+    [SAVE_PREFERRED_THEME]({ preferredTheme }, { theme, source }) {
+      switch (source) {
+        case 'os':
+          preferredTheme.fromOs = theme;
+          break;
+        case 'localstorage':
+          preferredTheme.fromLocalStorage = theme;
+          break;
+        case 'toggle':
+          preferredTheme.fromToggle = theme;
+          break;
+        default:
+          break;
+      }
+    },
+  },
+  actions: {
+    toggleTheme: ({ commit, getters }) => {
+      const theme = getters.isDarkTheme ? 'light' : 'dark';
+      commit(SAVE_PREFERRED_THEME, { theme, source: 'toggle' });
     },
   },
   getters: {
-    isDarkTheme: (state) => state.preferredTheme === 'dark',
+    isDarkTheme: ({ preferredTheme }) =>
+      preferredTheme.fromToggle === 'dark' ||
```

Comment:
alternative implementation
```js
[ 
  preferredTheme.fromToggle, 
  preferredTheme.fromLocalStorage, 
  preferredTheme.fromOs,
].filter(Boolean)[0] === 'dark'

Reply:
In FP it would look like so:
```js
// helpers
const pipe = (...fns) => (x) => fns.reduce((v, f) => f(v), x);
const filter = (fn, array) => array.filter(fn);
const head = array => array[0]
const eq = el1 => el2 => el1 === el2

// actual function
const isDarkTheme = pipe(
  filter(Boolean),
  head,
  eq('dark')
)

// usage
isDarkTheme([ 
  preferredTheme.fromToggle, 
  preferredTheme.fromLocalStorage, 
  preferredTheme.fromOs,
])
```

Reply:
@probil, is there a sense to use a FP library?

Reply:
FP makes code more declarative and less error prone, FP library would also provide all the helpers above

But in this particular example it's not necessary. You can stick with your approach or pick the first one I've mentioned.